File: .github/scripts/check-secret-rotations.py
#!/usr/bin/env python3
"""Open a GitHub issue when a long-lived secret is due for rotation.

The source of truth is the `rotations:` table in the front matter of
`docs/development/secret-rotations.md`. For each entry the script
computes (today - last-rotated). If that exceeds period-days minus
the reminder window (30 days), the script opens a single open issue
per secret with a stable title so reruns are idempotent.

The script does not close issues. After rotation the human edits the
`last-rotated` date in the doc, merges the change, and closes the
issue. The next scheduled run sees the new date and waits.

Exit codes:
- 0: ran cleanly (regardless of whether any issue was opened)
- 1: doc missing / malformed / `gh` failure / unknown secret kind
"""

from __future__ import annotations

import datetime as dt
import json
import os
import subprocess
import sys
from pathlib import Path

import yaml

REPO_ROOT = Path(__file__).resolve().parents[2]
ROTATION_DOC = REPO_ROOT / "docs" / "development" / "secret-rotations.md"
REMINDER_WINDOW_DAYS = 30
ISSUE_LABEL = "secret-rotation"
ASSIGNEE = "jeduden"


def _repo_url() -> str:
    """Return the absolute GitHub URL of this repository.

    GitHub Actions runners always populate GITHUB_SERVER_URL (e.g.
    https://github.com) and GITHUB_REPOSITORY (e.g. owner/name).
    Falling back to a hard-coded github.com URL keeps local runs
    (e.g. via `gh workflow run --workflow=...`) working when the
    envvars are absent.
    """
    server = os.environ.get("GITHUB_SERVER_URL", "https://github.com").rstrip("/")
    repo = os.environ.get("GITHUB_REPOSITORY", "jeduden/mdsmith")
    return f"{server}/{repo}"


def _front_matter(path: Path) -> dict:
    """Return the YAML front-matter block of a markdown file."""
    text = path.read_text(encoding="utf-8")
    if not text.startswith("---\n"):
        raise SystemExit(f"{path}: no front matter (must start with '---\\n')")
    end = text.find("\n---\n", 4)
    if end == -1:
        raise SystemExit(f"{path}: unterminated front matter")
    parsed = yaml.safe_load(text[4:end])
    if not isinstance(parsed, dict):
        raise SystemExit(f"{path}: front matter is not a mapping")
    return parsed


def _gh(args: list[str]) -> str:
    """Run `gh` and return stdout. Raise SystemExit on non-zero."""
    proc = subprocess.run(
        ["gh", *args], capture_output=True, text=True, check=False
    )
    if proc.returncode != 0:
        sys.stderr.write(
            f"gh {' '.join(args)} failed (exit {proc.returncode}):\n{proc.stderr}"
        )
        raise SystemExit(1)
    return proc.stdout


def _ensure_label() -> None:
    """Idempotently create the `secret-rotation` label.

    The first scheduled run of this script in a fresh repo would 422
    on `gh issue create --label ...` because the label does not yet
    exist. `gh label create --force` creates-or-updates the label so
    subsequent issue creation always succeeds. Subsequent runs are
    no-ops (the label just gets re-described, which is fine).
    """
    _gh(
        [
            "label",
            "create",
            ISSUE_LABEL,
            "--force",
            "--color",
            "C5DEF5",
            "--description",
            "Long-lived secret is due (or overdue) for rotation",
        ]
    )


def _existing_open_issue(title: str) -> int | None:
    """Return the number of an open issue whose title exactly matches."""
    out = _gh(
        [
            "issue",
            "list",
            "--state",
            "open",
            "--label",
            ISSUE_LABEL,
            "--json",
            "number,title",
            "--limit",
            "100",
        ]
    )
    for issue in json.loads(out or "[]"):
        if issue["title"] == title:
            return int(issue["number"])
    return None


def _due_state(today: dt.date, last_rotated: dt.date, period_days: int) -> tuple[str, int]:
    """Return ("ok"|"due"|"overdue", days_until_due).

    The reminder window is REMINDER_WINDOW_DAYS days before the period
    elapses. Anything past the period is overdue.
    """
    due_on = last_rotated + dt.timedelta(days=period_days)
    days = (due_on - today).days
    if days < 0:
        return "overdue", days
    if days <= REMINDER_WINDOW_DAYS:
        return "due", days
    return "ok", days


def _issue_body(entry: dict, status: str, days: int) -> str:
    """Compose the issue body. Wrapped in triple-quoted f-string."""
    name = entry["name"]
    last = entry["last-rotated"]
    period = entry["period-days"]
    provider = entry["provider"]
    issuer_url = entry["issuer-url"]
    used_by = entry["used-by"]
    scope = entry["scope"]
    if status == "overdue":
        headline = f"`{name}` is OVERDUE by {-days} days."
    else:
        headline = f"`{name}` is due in {days} days."
    repo_url = _repo_url()
    # Link to the doc without a heading anchor: GitHub's slug rules
    # (lowercase, hyphenated, with the em-dash stripped) do not match
    # what a naive name-to-slug derivation produces, so an anchored
    # link would 404 mid-page. The doc is short — a reader scrolling
    # to the `{name}` section finds it instantly.
    doc_url = f"{repo_url}/blob/main/docs/development/secret-rotations.md"
    workflow_url = f"{repo_url}/blob/main/.github/workflows/secret-rotation-reminder.yml"
    return f"""\
{headline}

| Field         | Value                                            |
|---------------|--------------------------------------------------|
| Provider      | {provider}                                       |
| Issuer URL    | <{issuer_url}>                                   |
| Used by       | {used_by}                                        |
| Scope         | {scope}                                          |
| Last rotated  | {last}                                           |
| Period (days) | {period}                                         |

Rotation procedure (jump to the `{name}` section):
{doc_url}

After rotation:

1. Update the `last-rotated` field for `{name}` in the front matter
   of `docs/development/secret-rotations.md`. Merge the change.
2. Close this issue.

This reminder is opened automatically by {workflow_url}.
"""


def main() -> int:
    fm = _front_matter(ROTATION_DOC)
    rotations = fm.get("rotations", [])
    if not isinstance(rotations, list) or not rotations:
        raise SystemExit(
            f"{ROTATION_DOC}: front matter has no `rotations:` list"
        )

    today = dt.date.today()
    opened: list[str] = []
    skipped: list[str] = []
    label_ensured = False

    for entry in rotations:
        for key in ("name", "last-rotated", "period-days", "provider", "issuer-url", "used-by", "scope"):
            if key not in entry:
                raise SystemExit(
                    f"rotation entry missing `{key}`: {entry!r}"
                )
        name = entry["name"]
        last_rotated = dt.date.fromisoformat(str(entry["last-rotated"]))
        period_days = int(entry["period-days"])
        status, days = _due_state(today, last_rotated, period_days)
        if status == "ok":
            continue
        title = f"Rotate {name} (last rotated {last_rotated.isoformat()})"
        if _existing_open_issue(title) is not None:
            skipped.append(name)
            continue
        if not label_ensured:
            # Bootstrap the label on first issue creation. Skipping
            # this would 422 on a fresh repo where the label has not
            # been created yet.
            _ensure_label()
            label_ensured = True
        body = _issue_body(entry, status, days)
        _gh(
            [
                "issue",
                "create",
                "--title",
                title,
                "--body",
                body,
                "--label",
                ISSUE_LABEL,
                "--assignee",
                ASSIGNEE,
            ]
        )
        opened.append(name)

    if opened:
        sys.stdout.write(
            f"opened secret-rotation reminders for: {', '.join(opened)}\n"
        )
    if skipped:
        sys.stdout.write(
            f"existing open reminders (skipped): {', '.join(skipped)}\n"
        )
    if not opened and not skipped:
        sys.stdout.write("no secrets due for rotation\n")
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: .github/scripts/check-secret-rotations.ts
#!/usr/bin/env bun
/**
 * Open a GitHub issue when a long-lived secret is due for rotation.
 *
 * Each tracked secret lives in its own file under
 * `docs/development/secret-rotations/`. The script globs that
 * directory, parses each file's YAML front matter, and computes
 * (today - lastRotated). If the elapsed time exceeds
 * `period-days` minus the reminder window (30 days), the script
 * opens a single labelled issue per file with a stable title so
 * reruns are idempotent.
 *
 * The script does not close issues directly. The reminder issue
 * body points the maintainer at the `Record Secret Rotation`
 * workflow, which opens a PR that updates `lastRotated` and
 * includes `Closes #N`. Merging that PR records the rotation date
 * AND closes the reminder in one step. The next scheduled run
 * sees the new date and stays quiet until the next expiry window.
 *
 * Exit codes:
 * - 0: ran cleanly (regardless of whether any issue was opened)
 * - 1: doc missing / malformed / `gh` failure / unknown entry kind
 */

import { $, Glob } from "bun";
import { basename, resolve } from "node:path";
import { parse as yamlParse } from "yaml";

const REPO_ROOT = resolve(import.meta.dir, "..", "..");
const ROTATIONS_DIR = resolve(REPO_ROOT, "docs/development/secret-rotations");
const REMINDER_WINDOW_DAYS = 30;
const ISSUE_LABEL = "secret-rotation";
// Derive the assignee from GITHUB_REPOSITORY_OWNER (auto-set by
// GitHub Actions) so an org/owner change doesn't silently break
// the workflow. Falls back to a sensible default for local runs
// outside Actions.
const ASSIGNEE = process.env.GITHUB_REPOSITORY_OWNER || "jeduden";

interface RotationEntry {
  title: string;
  lastRotated: string;
  periodDays: number;
  provider: string;
  issuerUrl: string;
  usedBy: string;
  scope: string;
}

export class SystemExit extends Error {}

/** Validate that `s` is a real calendar date in YYYY-MM-DD form.
 * Round-trips the parsed components so normalized invalid dates
 * (e.g. `2026-02-31` parsing to March 3) are rejected. */
export function isIsoDate(s: string): boolean {
  if (!/^\d{4}-\d{2}-\d{2}$/.test(s)) return false;
  const parsed = new Date(`${s}T00:00:00Z`);
  if (Number.isNaN(parsed.getTime())) return false;
  const year = String(parsed.getUTCFullYear()).padStart(4, "0");
  const month = String(parsed.getUTCMonth() + 1).padStart(2, "0");
  const day = String(parsed.getUTCDate()).padStart(2, "0");
  return `${year}-${month}-${day}` === s;
}

/** Compute today's date in UTC. The cron schedule is UTC, so the
 * computed due-state must match the workflow's wall clock. */
export function utcToday(): Date {
  const now = new Date();
  return new Date(Date.UTC(
    now.getUTCFullYear(),
    now.getUTCMonth(),
    now.getUTCDate(),
  ));
}

/** Days between two UTC midnights, truncated to integer. */
export function daysBetween(later: Date, earlier: Date): number {
  const msPerDay = 24 * 60 * 60 * 1000;
  return Math.floor((later.getTime() - earlier.getTime()) / msPerDay);
}

/** Extract the YAML front matter block from a markdown file. */
export function parseFrontMatter(text: string, path: string): Record<string, unknown> {
  if (!text.startsWith("---\n")) {
    throw new SystemExit(`${path}: no front matter (must start with '---\\n')`);
  }
  const end = text.indexOf("\n---\n", 4);
  if (end === -1) {
    throw new SystemExit(`${path}: unterminated front matter`);
  }
  const parsed = yamlParse(text.slice(4, end));
  if (parsed === null || typeof parsed !== "object" || Array.isArray(parsed)) {
    throw new SystemExit(`${path}: front matter is not a mapping`);
  }
  return parsed as Record<string, unknown>;
}

export type DueState = "ok" | "due" | "overdue";

/** A signed days-until-due. Positive while the rotation is still
 * in the future, zero on the due date, negative once past due.
 * Callers format the value differently per status: "due in N
 * days" / "due today" / "OVERDUE by N days" (negate). */
export interface DueResult {
  status: DueState;
  daysUntilDue: number;
}

export function dueState(today: Date, lastRotated: Date, periodDays: number): DueResult {
  const dueOn = new Date(lastRotated);
  dueOn.setUTCDate(dueOn.getUTCDate() + periodDays);
  const daysUntilDue = daysBetween(dueOn, today);
  if (daysUntilDue < 0) return { status: "overdue", daysUntilDue };
  if (daysUntilDue <= REMINDER_WINDOW_DAYS) return { status: "due", daysUntilDue };
  return { status: "ok", daysUntilDue };
}

/** Return the absolute GitHub URL of this repository. */
export function repoUrl(): string {
  const server = (process.env.GITHUB_SERVER_URL ?? "https://github.com").replace(/\/+$/, "");
  const repo = process.env.GITHUB_REPOSITORY ?? "jeduden/mdsmith";
  return `${server}/${repo}`;
}

interface OpenIssue { number: number; title: string }

async function existingOpenIssue(title: string): Promise<number | null> {
  // Narrow the candidate set server-side via GitHub search.
  // `in:title "<phrase>"` matches issues whose title contains the
  // quoted phrase, so a repo with hundreds of unrelated
  // secret-rotation-labelled issues never pushes the real match
  // past the --limit ceiling. The exact-string check below catches
  // GitHub search's tokenized/fuzzy behavior — we only treat the
  // issue as existing if its title is byte-for-byte identical.
  const out = await $`gh issue list \
    --state open \
    --label ${ISSUE_LABEL} \
    --search ${`in:title "${title.replace(/"/g, "")}"`} \
    --json number,title \
    --limit 100`.text();
  const issues = JSON.parse(out || "[]") as OpenIssue[];
  for (const issue of issues) {
    if (issue.title === title) return issue.number;
  }
  return null;
}

let labelEnsured = false;
async function ensureLabel(): Promise<void> {
  if (labelEnsured) return;
  await $`gh label create ${ISSUE_LABEL} \
    --force \
    --color C5DEF5 \
    --description ${"Long-lived secret is due (or overdue) for rotation"}`.quiet();
  labelEnsured = true;
}

export function issueBody(entry: RotationEntry, fileBasename: string, due: DueResult): string {
  let headline: string;
  if (due.status === "overdue") {
    // daysUntilDue is negative once past due; negate for the
    // English "OVERDUE by N days" reading.
    headline = `\`${entry.title}\` is OVERDUE by ${-due.daysUntilDue} days.`;
  } else if (due.daysUntilDue === 0) {
    headline = `\`${entry.title}\` is due today.`;
  } else {
    headline = `\`${entry.title}\` is due in ${due.daysUntilDue} days.`;
  }
  const fileUrl = `${repoUrl()}/blob/main/docs/development/secret-rotations/${fileBasename}`;
  const reminderUrl = `${repoUrl()}/blob/main/.github/workflows/secret-rotation-reminder.yml`;
  const recordUrl = `${repoUrl()}/actions/workflows/record-secret-rotation.yml`;
  // Table padding is intentionally minimal: GitHub-flavored
  // markdown collapses extra whitespace, so visually aligning
  // columns in the source has no effect on the rendered issue
  // and just creates noisy diffs whenever a value's length
  // changes.
  return [
    `${headline}`,
    ``,
    `| Field | Value |`,
    `|---|---|`,
    `| Provider | ${entry.provider} |`,
    `| Issuer URL | <${entry.issuerUrl}> |`,
    `| Used by | ${entry.usedBy} |`,
    `| Scope | ${entry.scope} |`,
    `| lastRotated | ${entry.lastRotated} |`,
    `| Period (days) | ${entry.periodDays} |`,
    ``,
    `Rotation procedure:`,
    `${fileUrl}`,
    ``,
    `After rotating the credential at the issuer, do not`,
    `hand-edit the front matter or close this issue.`,
    `Instead, run the **Record Secret Rotation** workflow:`,
    `${recordUrl}`,
    ``,
    `Pick \`${entry.title}\` from the dropdown and click \`Run workflow\`.`,
    `The workflow opens a PR that updates \`lastRotated\``,
    `and includes \`Closes #\` referencing this issue, so`,
    `the merge both records the rotation and closes this`,
    `reminder in one step.`,
    ``,
    `This reminder was opened automatically by ${reminderUrl}.`,
    "",
  ].join("\n");
}

const REQUIRED_FRONT_MATTER_KEYS: (keyof RotationEntry)[] = [
  "title", "lastRotated", "periodDays", "provider", "issuerUrl", "usedBy", "scope",
];

/** Validate a single rotation entry's front matter and project it
 * into a `RotationEntry`. Throws `SystemExit` with a path-prefixed
 * message on any structural failure. Separated from `loadRotations`
 * so the validation rules are reachable from unit tests without
 * filesystem fixtures. */
export function validateRotationEntry(fm: Record<string, unknown>, path: string): RotationEntry {
  for (const key of REQUIRED_FRONT_MATTER_KEYS) {
    if (!(key in fm)) {
      throw new SystemExit(`${path}: front matter missing \`${key}\``);
    }
  }
  const lastStr = String(fm.lastRotated);
  if (!isIsoDate(lastStr)) {
    throw new SystemExit(
      `${path}: \`lastRotated\` is not a valid ISO-8601 date (${JSON.stringify(lastStr)})`,
    );
  }
  const periodDays = Number(fm.periodDays);
  if (!Number.isInteger(periodDays)) {
    throw new SystemExit(
      `${path}: \`periodDays\` is not an integer (${JSON.stringify(fm.periodDays)})`,
    );
  }
  // A zero or negative period would compute a due date on or
  // before lastRotated, so every run would treat the secret as
  // overdue and the reminder workflow would never go quiet.
  // Reject the value at load time with a clear pointer to the
  // bad file rather than silently spamming issues.
  if (periodDays <= 0) {
    throw new SystemExit(
      `${path}: \`periodDays\` must be a positive integer (got ${periodDays})`,
    );
  }
  return {
    title: String(fm.title),
    lastRotated: lastStr,
    periodDays,
    provider: String(fm.provider),
    issuerUrl: String(fm.issuerUrl),
    usedBy: String(fm.usedBy),
    scope: String(fm.scope),
  };
}

/** Load every per-secret rotation file from ROTATIONS_DIR. */
async function loadRotations(): Promise<{ entry: RotationEntry; fileBasename: string }[]> {
  const glob = new Glob("*.md");
  const entries: { entry: RotationEntry; fileBasename: string }[] = [];
  for await (const rel of glob.scan({ cwd: ROTATIONS_DIR })) {
    const fileBasename = basename(rel);
    const path = resolve(ROTATIONS_DIR, rel);
    const text = await Bun.file(path).text();
    const fm = parseFrontMatter(text, path);
    entries.push({ entry: validateRotationEntry(fm, path), fileBasename });
  }
  // Sort for deterministic iteration order regardless of FS ordering.
  entries.sort((a, b) => a.entry.title.localeCompare(b.entry.title));
  return entries;
}

async function main(): Promise<number> {
  const rotations = await loadRotations();
  if (rotations.length === 0) {
    throw new SystemExit(`${ROTATIONS_DIR}: no per-secret files found`);
  }

  const today = utcToday();
  const opened: string[] = [];
  const skipped: string[] = [];

  for (const { entry, fileBasename } of rotations) {
    const lastRotated = new Date(`${entry.lastRotated}T00:00:00Z`);
    const due = dueState(today, lastRotated, entry.periodDays);
    if (due.status === "ok") continue;
    const title = `Rotate ${entry.title} (lastRotated ${entry.lastRotated})`;
    if (await existingOpenIssue(title) !== null) {
      skipped.push(entry.title);
      continue;
    }
    await ensureLabel();
    const body = issueBody(entry, fileBasename, due);
    await $`gh issue create \
      --title ${title} \
      --body ${body} \
      --label ${ISSUE_LABEL} \
      --assignee ${ASSIGNEE}`.quiet();
    opened.push(entry.title);
  }

  if (opened.length) {
    console.log(`opened secret-rotation reminders for: ${opened.join(", ")}`);
  }
  if (skipped.length) {
    console.log(`existing open reminders (skipped): ${skipped.join(", ")}`);
  }
  if (!opened.length && !skipped.length) {
    console.log("no secrets due for rotation");
  }
  return 0;
}

// Gate the auto-run on `import.meta.main` so test files that
// import this module for its pure exports do not also fire the
// shell-out paths in `main()` as a side effect of the import.
if (import.meta.main) {
  try {
    process.exit(await main());
  } catch (err) {
    if (err instanceof SystemExit) {
      process.stderr.write(`${err.message}\n`);
      process.exit(1);
    }
    throw err;
  }
}
